draft-irtf-cfrg-bls-signature.md

```diff
@@ -0,0 +1,1290 @@
+
+        signature\_subgroup\_check(ostr) := subgroup\_check\_E1(ostr)
+
+  - For minimal-pubkey-size:
+
+        pubkey\_subgroup\_check(ostr) := subgroup\_check\_E1(ostr)
+
+        signature\_subgroup\_check(ostr) := subgroup\_check\_E2(ostr)
+
+##  KeyGen {#keygen}
+
+The KeyGen algorithm generates a pair (PK, SK) deterministically using
+the secret octet string IKM.
+
+KeyGen uses HKDF [@!RFC5869] instantiated with the hash function H.
+
+For security, IKM MUST be infeasible to guess, e.g.,
+generated by a trusted source of randomness.
+IKM MUST be at least 32 bytes long, but it MAY be longer.
+
+Because KeyGen is deterministic, implementations MAY choose either to
+store the resulting (PK, SK) or to store IKM and call KeyGen to derive
+the keys when necessary.
+
+~~~
+(PK, SK) = KeyGen(IKM)
+
+Inputs:
+- IKM, a secret octet string. See requirements above.
+
+Outputs:
+- PK, a public key encoded as an octet string.
+- SK, the corresponding secret key, an integer 0 <= SK < r.
+
+Definitions:
+- HKDF-Extract is as defined in RFC5869, instantiated with hash H.
+- HKDF-Expand is as defined in RFC5869, instantiated with hash H.
+- L is the integer given by ceil((1.5 * ceil(log2(r))) / 8).
+- "BLS-SIG-KEYGEN-SALT-" is an ASCII string comprising 20 octets.
+- "" is the empty string.
+
+Procedure:
+1. PRK = HKDF-Extract("BLS-SIG-KEYGEN-SALT-", IKM)
+2. OKM = HKDF-Expand(PRK, "", L)
+3. x = OS2IP(OKM) mod r
+4. xP = x * P
+5. SK = x
+6. PK = point_to_pubkey(xP)
+7. return (PK, SK)
+~~~
+
+## KeyValidate {#keyvalidate}
+
+The KeyValidate algorithm ensures that a public key is valid.
+
+~~~
+result = KeyValidate(PK)
+
+Inputs:
+- PK, a public key in the format output by KeyGen.
+
+Outputs:
+- result, either VALID or INVALID
+
+Procedure:
+1. xP = pubkey_to_point(PK)
+2. If xP is INVALID, return INVALID
+3. If pubkey_subgroup_check(xP) is INVALID, return INVALID
+4. return VALID
+~~~
+
+## CoreSign {#coresign}
+
+The CoreSign algorithm computes a signature from SK, a secret key,
+and message, an octet string.
+
+~~~
+signature = CoreSign(SK, message)
+
+Inputs:
+- SK, the secret key output by an invocation of KeyGen.
+- message, an octet string.
+
+Outputs:
+- signature, an octet string.
+
+Procedure:
+1. Q = hash_to_point(message)
+2. R = SK * Q
+3. signature = point_to_signature(R)
+4. return signature
+~~~
+
+## CoreVerify {#coreverify}
+
+The CoreVerify algorithm checks that a signature is valid for
+the octet string message under the public key PK.
+
+The public key PK must satisfy KeyValidate(PK) == VALID ((#keyvalidate)).
+
+~~~
+result = CoreVerify(PK, message, signature)
+
+Inputs:
+- PK, a public key in the format output by KeyGen.
+- message, an octet string.
+- signature, an octet string in the format output by CoreSign.
+
+Outputs:
+- result, either VALID or INVALID.
+
+Procedure:
+1. R = signature_to_point(signature)
+2. If R is INVALID, return INVALID
+3. If signature_subgroup_check(R) is INVALID, return INVALID
```

Comment:
Here, the input R is a point, whereas signature\_subgroup\_check(ostr) takes octet string as input.

I'm leaning towards not defining the {signature, pubkey}_subgroup_check routines because they create an additional level of indirection. Rather, just define subgroup_check that takes a point (which also determines E1 / E2) and checks whether it lies in the subgroup. That said, I understand we may want to specify E1/E2 more explicitly.

Reply:
Oh! Right. Actually, I think that's a copypaste error in the "Parameters" subsection of "Core operations" --- the definition of subgroup_check in section 1 says that it takes a point, not an octet string.

Reply:
OK, fixed signature_subgroup_check and pubkey_subgroup_check to agree with the definition of subgroup_check from section 1.

Reply:
Regarding the indirection:

I don't totally love the indirection either---really, I don't love it for any of these functions. But it seems like leaving it implicit is much less clear for someone who doesn't already know exactly what's going on. Maybe I'm worrying about this too much.

Reply:
Also, just as an aesthetic matter, it seems like if we get rid of the subgroup_check indirection, we should get rid of all of the indirections, right?